src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -544,7 +544,21 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
             {
                 if (constantValueOpt == null)
                 {
-                    diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
+                    if (convertedExpression is BoundConversion conversion && conversion.ConversionKind == ConversionKind.ImplicitUserDefined)
+                    {
+                        if (expression.IsLiteralNull() && !inputType.IsNullableType())
+                        {
+                            diagnostics.Add(ErrorCode.ERR_ValueCantBeNull, patternExpression.Location, inputType);
+                        }
+                        else
+                        {
+                            diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, expression.Type!, inputType);
```

Comment:
```suggestion
                            Debug.Assert(expression.Type is not null);
                            diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, expression.Type, inputType);
```

Prefer debug assertions over `!` suppressions.